std/string.d

```diff
@@ -2937,26 +2937,96 @@ if (isForwardRange!Range && isSomeChar!(ElementEncodingType!Range) &&
 {
     static import std.ascii;
     static import std.uni;
-    import std.utf : decodeFront;
 
-    while (!input.empty)
+    static if (is(Unqual!(ElementEncodingType!Range) == dchar)
+        || is(Unqual!(ElementEncodingType!Range) == wchar))
     {
-        auto c = input.front;
-        if (std.ascii.isASCII(c))
+        // Decoding is never needed for dchar. It happens not to be needed
+        // here for wchar because no whitepace is outside the basic
+        // multilingual plane meaning every whitespace character is encoded
+        // with a single wchar and due to the design of UTF-16 those wchars
+        // will not occur as part of the encoding of multi-wchar codepoints.
+        static if (isDynamicArray!Range || (isRandomAccessRange!Range && hasLength!Range
+            && __traits(compiles, { size_t s = input.length / 2; input = input[s .. $]; })))
```

Comment:
I don't think the traits compiles thing is necessary. Just use `hasSlicing`.

Reply:
I need to also check that the result of the slice can be assigned back to input.

Reply:
`hasSlicing` implies that `input[x .. y]` returns `typeof(input)`. All you have to do is check that `input` is assignable to `input`.

Reply:
With your rewrite I don't need to check even that.